kv/kvserver: TestReplicaStateMachineChangeReplicas failed

kv/kvserver.TestReplicaStateMachineChangeReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7620213?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7620213?buildTab=artifacts#/) on master @ [f0554bc215e31ff53e644936ba645bd895ec0d15](https://github.com/cockroachdb/cockroach/commits/f0554bc215e31ff53e644936ba645bd895ec0d15):


```
=== RUN   TestReplicaStateMachineChangeReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaStateMachineChangeReplicas2621667056
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=false
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=true
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x16b6c08]

goroutine 110308098 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc013e0b170, {0x5ca2b80, 0xc01ab58420})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:248 +0xa5
panic({0x4308020, 0x82658e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).getBreaker(0xc013dbfd00, 0x2, 0x0?)
	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:274 +0x88
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).GetCircuitBreaker(...)
	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:267
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).SendAsync(0xc009db4c40, 0xc013a3c4e0, 0xa5?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:513 +0xe6
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRequest(0xc00def8400, {0x5ca2b80, 0xc015798f90}, 0xc013a3c4e0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1566 +0xd9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRaftMuLocked(_, {_, _}, {0x3, 0x2, 0x1, 0x6, 0x6, 0xd, {0xc01194c5a0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1535 +0x4be
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessagesRaftMuLocked(0xc00def8400, {0x5ca2b80, 0xc015798f90}, {0xc013f0efc0?, 0x1, 0xe?}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1473 +0x270
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:917 +0x1137
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:664 +0x170
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc0109ba000, 0xc017400fc0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:635 +0xb8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker(0xc00e832f00, {0x5ca2b80, 0xc01ab58420})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:333 +0x1ea
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:489 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:480 +0x43b
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #91047 kv/kvserver: TestReplicaStateMachineChangeReplicas failed [C-test-failure O-robot T-kv branch-release-22.1]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaStateMachineChangeReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21673

kv/kvserver.TestReplicaStateMachineChangeReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7854063?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7854063?buildTab=artifacts#/) on master @ [146556e19f5e4fdc8c3e6a623b280cc33aee4d18](https://github.com/cockroachdb/cockroach/commits/146556e19f5e4fdc8c3e6a623b280cc33aee4d18):


```
=== RUN   TestReplicaStateMachineChangeReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaStateMachineChangeReplicas34986891
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=false
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=true
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x16d5ec8]

goroutine 113106868 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0016ebc68?, {0x5d13c00, 0xc0109edb90})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x4359f40, 0x830d900})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).getBreaker(0xc01abcfc00, 0x2, 0x0?)
	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:274 +0x88
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).GetCircuitBreaker(...)
	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:267
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).SendAsync(0xc00da647e0, 0xc0017829c0, 0xc5?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:513 +0xe6
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRequest(0xc001844c00, {0x5d13c00, 0xc0142d5c80}, 0xc0017829c0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1567 +0xd9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRaftMuLocked(_, {_, _}, {0x3, 0x2, 0x1, 0x6, 0x6, 0xd, {0xc010f91710, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1536 +0x4be
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessagesRaftMuLocked(0xc001844c00, {0x5d13c00, 0xc0142d5c80}, {0xc0125517a0?, 0x1, 0xe?}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1474 +0x270
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:917 +0x1137
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:664 +0x170
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc00a33a000, 0xc0142d41e0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:634 +0xb8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker(0xc00ea7a0f0, {0x5d13c00, 0xc0109edb90})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:333 +0x1ea
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92554 kv/kvserver: TestReplicaStateMachineChangeReplicas failed [C-test-failure O-robot T-kv branch-release-22.2]
- #91047 kv/kvserver: TestReplicaStateMachineChangeReplicas failed [C-test-failure O-robot T-kv-replication branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaStateMachineChangeReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestReplicaStateMachineChangeReplicas [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7882111?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7882111?buildTab=artifacts#/) on master @ [ec095bc2fdbe4e518b076db20e4920fab67222bf](https://github.com/cockroachdb/cockroach/commits/ec095bc2fdbe4e518b076db20e4920fab67222bf):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x16de428]

```
Stack: 

```
goroutine 110110547 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0025b7f30?, {0x5d4a4e0, 0xc01e801860})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x43804e0, 0x83648f0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).getBreaker(0xc014df9780, 0x2, 0x0?)
	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:274 +0x88
github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).GetCircuitBreaker(...)
	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:267
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).SendAsync(0xc009dd8620, 0xc00d292ea0, 0xe5?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:513 +0xe6
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRequest(0xc00ba10000, {0x5d4a4e0, 0xc00fdbe180}, 0xc00d292ea0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1567 +0xd9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessageRaftMuLocked(0xc00ba10000, {0x5d4a4e0, 0xc00fdbe180}, {0x3, 0x2, 0x1, 0x6, 0x6, 0xd, {0xc00c125ef0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1536 +0x4b9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendRaftMessagesRaftMuLocked(0xc00ba10000, {0x5d4a4e0, 0xc00fdbe180}, {0xc01c04a900?, 0x1, 0xe?}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1474 +0x217
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:917 +0x1137
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:664 +0x170
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc0070ce000, 0xc00d27b4a0?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:634 +0xb8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker(0xc015bcf4a0, {0x5d4a4e0, 0xc01e801860})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:333 +0x1ea
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestReplicaStateMachineChangeReplicas
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaStateMachineChangeReplicas1312279805
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=false
=== RUN   TestReplicaStateMachineChangeReplicas/add_replica=true
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92554 kv/kvserver: TestReplicaStateMachineChangeReplicas failed [C-test-failure O-robot T-kv branch-release-22.2]
- #91047 kv/kvserver: TestReplicaStateMachineChangeReplicas failed [C-test-failure O-robot T-kv-replication branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaStateMachineChangeReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_application_state_machine_test.go
+++ b/pkg/kv/kvserver/replica_application_state_machine_test.go
@@ -145,6 +145,14 @@ func TestReplicaStateMachineChangeReplicas(t *testing.T) {
 			require.Error(t, err)
 			require.IsType(t, &roachpb.RangeNotFoundError{}, err)
 		}
+		// Set a destroyStatus to make sure there won't be any raft processing once
+		// we release raftMu. We applied a command but not one from the raft log, so
+		// should there be a command in the raft log (i.e. some errant lease request
+		// or whatnot) this will fire assertions because it will conflict with the
+		// log index that we pulled out of thin air above.
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.mu.destroyStatus.Set(errors.New("test done"), destroyReasonRemoved)
 	})
 }
 

--- a/pkg/kv/kvserver/replica_command.go
+++ b/pkg/kv/kvserver/replica_command.go
@@ -658,26 +658,31 @@ func (r *Replica) AdminMerge(
 			return errors.Errorf("ranges not collocated; %s != %s", lReplicas, rReplicas)
 		}
 
-		// Ensure that every current replica of the LHS has been initialized.
-		// Otherwise there is a rare race where the replica GC queue can GC a
-		// replica of the RHS too early. The comment on
-		// TestStoreRangeMergeUninitializedLHSFollower explains the situation in full.
-		if err := waitForReplicasInit(
-			ctx, r.store.cfg.NodeDialer, origLeftDesc.RangeID, origLeftDesc.Replicas().Descriptors(),
-		); err != nil {
-			return errors.Wrap(err, "waiting for all left-hand replicas to initialize")
-		}
-		// Out of an abundance of caution, also ensure that replicas of the RHS have
-		// all been initialized. If for whatever reason the initial upreplication
-		// snapshot for a NON_VOTER on the RHS fails, it will have to get picked up
-		// by the raft snapshot queue to upreplicate and may be uninitialized at
-		// this point. As such, if we send a subsume request to the RHS in this sort
-		// of state, we will wastefully and unintentionally block all traffic on it
-		// for 5 seconds.
-		if err := waitForReplicasInit(
-			ctx, r.store.cfg.NodeDialer, rightDesc.RangeID, rightDesc.Replicas().Descriptors(),
-		); err != nil {
-			return errors.Wrap(err, "waiting for all right-hand replicas to initialize")
+		disableWaitForReplicasInTesting := r.store.TestingKnobs() != nil &&
+			r.store.TestingKnobs().DisableMergeWaitForReplicasInit
+
+		if !disableWaitForReplicasInTesting {
+			// Ensure that every current replica of the LHS has been initialized.
+			// Otherwise there is a rare race where the replica GC queue can GC a
+			// replica of the RHS too early. The comment on
+			// TestStoreRangeMergeUninitializedLHSFollower explains the situation in full.
+			if err := waitForReplicasInit(
+				ctx, r.store.cfg.NodeDialer, origLeftDesc.RangeID, origLeftDesc.Replicas().Descriptors(),
+			); err != nil {
+				return errors.Wrap(err, "waiting for all left-hand replicas to initialize")
+			}
+			// Out of an abundance of caution, also ensure that replicas of the RHS have
+			// all been initialized. If for whatever reason the initial upreplication
+			// snapshot for a NON_VOTER on the RHS fails, it will have to get picked up
+			// by the raft snapshot queue to upreplicate and may be uninitialized at
+			// this point. As such, if we send a subsume request to the RHS in this sort
+			// of state, we will wastefully and unintentionally block all traffic on it
+			// for 5 seconds.
+			if err := waitForReplicasInit(
+				ctx, r.store.cfg.NodeDialer, rightDesc.RangeID, rightDesc.Replicas().Descriptors(),
+			); err != nil {
+				return errors.Wrap(err, "waiting for all right-hand replicas to initialize")
+			}
 		}
 
 		mergeReplicas := lReplicas.Descriptors()
@@ -750,6 +755,9 @@ func (r *Replica) AdminMerge(
 
 		err = contextutil.RunWithTimeout(ctx, "waiting for merge application", mergeApplicationTimeout,
 			func(ctx context.Context) error {
+				if disableWaitForReplicasInTesting {
+					return nil
+				}
 				return waitForApplication(ctx, r.store.cfg.NodeDialer, rightDesc.RangeID, mergeReplicas,
 					rhsSnapshotRes.LeaseAppliedIndex)
 			})
@@ -822,11 +830,6 @@ func waitForApplication(
 	replicas []roachpb.ReplicaDescriptor,
 	leaseIndex uint64,
 ) error {
-	if dialer == nil && len(replicas) == 1 {
-		// This early return supports unit tests (testContext{}) that also
-		// want to perform merges.
-		return nil
-	}
 	g := ctxgroup.WithContext(ctx)
 	for _, repl := range replicas {
 		repl := repl // copy for goroutine
@@ -856,11 +859,6 @@ func waitForReplicasInit(
 	rangeID roachpb.RangeID,
 	replicas []roachpb.ReplicaDescriptor,
 ) error {
-	if dialer == nil && len(replicas) == 1 {
-		// This early return supports unit tests (testContext{}) that also
-		// want to perform merges.
-		return nil
-	}
 	return contextutil.RunWithTimeout(ctx, "wait for replicas init", 5*time.Second, func(ctx context.Context) error {
 		g := ctxgroup.WithContext(ctx)
 		for _, repl := range replicas {

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -162,6 +162,7 @@ func (tc *testContext) Start(ctx context.Context, t testing.TB, stopper *stop.St
 	// testContext tests like to move the manual clock around and assume that they can write at past
 	// timestamps.
 	cfg.TestingKnobs.DontCloseTimestamps = true
+	cfg.TestingKnobs.DisableMergeWaitForReplicasInit = true
 	tc.StartWithStoreConfig(ctx, t, stopper, cfg)
 }
 
@@ -14008,7 +14009,7 @@ func TestStoreTenantMetricsAndRateLimiterRefcount(t *testing.T) {
 			Key: leftRepl.Desc().StartKey.AsRawKey(),
 		},
 	}, "testing")
-	require.Nil(t, pErr)
+	require.NoError(t, pErr.GoError())
 
 	// The store metrics no longer track tenant 123.
 	require.Equal(t,

--- a/pkg/kv/kvserver/store_test.go
+++ b/pkg/kv/kvserver/store_test.go
@@ -207,7 +207,15 @@ func createTestStoreWithoutStart(
 	eng := storage.NewDefaultInMemForTesting()
 	stopper.AddCloser(eng)
 	require.Nil(t, cfg.Transport)
-	cfg.Transport = NewDummyRaftTransport(cfg.Settings, cfg.AmbientCtx.Tracer)
+
+	require.NotNil(t, cfg.Gossip) // was set above already
+	// Even though testContext is fundamentally a single-store test, some tests
+	// will try config changes, etc, so we will see some use of the transport
+	// and it's important that this doesn't cause crashes. Just set up the
+	// "real thing" since it's straightforward enough.
+	cfg.NodeDialer = nodedialer.New(rpcContext, gossip.AddressResolver(cfg.Gossip))
+	cfg.Transport = NewRaftTransport(cfg.AmbientCtx, cfg.Settings, cfg.Tracer(), cfg.NodeDialer, server, stopper)
+
 	stores := NewStores(cfg.AmbientCtx, cfg.Clock)
 	nodeDesc := &roachpb.NodeDescriptor{NodeID: 1}
 

--- a/pkg/kv/kvserver/testing_knobs.go
+++ b/pkg/kv/kvserver/testing_knobs.go
@@ -473,6 +473,10 @@ type StoreTestingKnobs struct {
 	// Replica.executeReadOnlyBatch after checks have successfully determined
 	// execution can proceed but a storage snapshot has not been acquired.
 	PreStorageSnapshotButChecksCompleteInterceptor func(replica *Replica)
+
+	// DisableMergeWaitForReplicasInit skips the waitForReplicasInit calls
+	// during merges. Useful for testContext tests that want to use merges.
+	DisableMergeWaitForReplicasInit bool
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

